pkg/datapath/linux/devices_controller.go

```diff
@@ -0,0 +1,531 @@
+
+		case <-timer.C:
+			timerStopped = true
+			if dc.processBatch(batch) {
+				batch = map[int][]any{}
+			}
+		}
+	}
+}
+
+func deviceAddressFromAddrUpdate(upd netlink.AddrUpdate) tables.DeviceAddress {
+	return tables.DeviceAddress{
+		Addr:  ip.MustAddrFromIP(upd.LinkAddress.IP),
+		Flags: upd.Flags,
+		Scope: upd.Scope,
+	}
+}
+
+// processBatch processes a batch of link, route and address updates. If it successfully
+// commits it to the state it returns true, if the batch is incomplete (e.g. route updates
+// for non-existing link), it returns false and the batch should be expanded with newer
+// updates and tried again.
+func (dc *devicesController) processBatch(batch map[int][]any) (ok bool) {
+	txn := dc.DB.WriteTxn()
+	devicesWriter := dc.DeviceTable.Writer(txn)
+	routesWriter := dc.RouteTable.Writer(txn)
+
+	// Process link updates first so we can access the link metadata when processing
+	// addresses and routes.
+	for index, updates := range batch {
+		for _, u := range updates {
+			if l, ok := u.(netlink.LinkUpdate); ok {
+				if dc.shouldSkipDevice(l) {
+					delete(batch, index)
+					continue
+				}
+
+				if l.Header.Type == unix.RTM_DELLINK {
+					delete(dc.links, index)
+					delete(batch, index)
+					devicesWriter.DeleteAll(tables.DeviceByIndex(index))
+					routesWriter.DeleteAll(tables.RouteByLinkIndex(index))
+					break
+				} else {
+					dc.links[index] = l.Link
+				}
+			}
+		}
+	}
+
+	for index, updates := range batch {
+		link := dc.links[index]
+		if link == nil {
+			// Update for a link we have not yet observed, abort and try again
+			// later.
+			dc.Log.Debugf("link %d not found, retrying later", index)
+			txn.Abort()
+			return false
+		}
+
+		name := link.Attrs().Name
+
+		d, _ := devicesWriter.First(tables.DeviceByIndex(index))
+		if d == nil {
+			d = &tables.Device{
+				Index: index,
+				Name:  name,
+				Link:  link,
+			}
+		} else {
+			d = d.DeepCopy()
+		}
+
+		for _, u := range updates {
+			switch u := u.(type) {
+			case netlink.AddrUpdate:
+				if u.Scope == unix.RT_SCOPE_LINK {
+					// Ignore link local addresses.
+					continue
+				}
+				addr := deviceAddressFromAddrUpdate(u)
+				if u.NewAddr {
+					d.Addrs = append(d.Addrs, addr)
+				} else {
+					i := slices.Index(d.Addrs, addr)
+					if i >= 0 {
+						d.Addrs = slices.Delete(d.Addrs, i, i+1)
+					}
+				}
+			case netlink.RouteUpdate:
+				r := tables.Route{
+					Table:     u.Table,
+					LinkName:  name,
+					LinkIndex: index,
+					Scope:     uint8(u.Scope),
+					Dst:       u.Dst,
+					Src:       u.Src,
+					Gw:        u.Gw,
+				}
+				if u.Type == unix.RTM_NEWROUTE {
+					if err := routesWriter.Insert(&r); err != nil {
+						log.WithError(err).Error("Insert into routes table failed")
+					}
+				} else {
+					routesWriter.Delete(&r)
+				}
+			}
+		}
+		// Sort the addresses so that secondary addresses are at bottom and then
+		// sorted by scope.
+		sort.SliceStable(d.Addrs, func(i, j int) bool {
```

Comment:
This sorting is only needed to pick the address for kube-proxy replacement, but that's not yet implemented. I removed the sorting from here and will add it directly to the function that picks the NodePort frontend address (once KPR gets ported to use devices table instead of the pkg/node globals).